### PR TITLE
refine ownership model for ServiceContext

### PR DESCRIPTION
**Pull request by agavra**
### Description 
This patch refines the ownership model for ServiceContext so that only `KafkaRestApplication` and `KsqlEngine` hold onto a reference to a ServiceContext beyond a local variable or parameter (as instance field). This sets us up to inject modified ServiceContexts as dependencies without having to track down ownership leaks.

There is one other small change: we create an `ExecutionContext` on each execution instead of holding one primary one around. This clarifies the distinction between `ServiceContext` (created once per service) and `ExecutionContext` (created once per execution).

### Testing done 
- Existing unit test coverage for modified code is high

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by agavra** [comment]
**tl;dr** we're closing out this and #2396 in favor of doing more topic management on the REST endpoint side of things.

Here's a summary of an offline conversation: 

- Motivation for this PR is to be able to pass in mock ServiceContexts for recovery mode (to avoid topic creations/deletes) 
- We both agree that the ServiceContext should be owned by the application, not the engine. The reason this PR doesn't do that is because it's blocked on #2396. 
- Passing in `ServiceContext` each time becomes a little unwieldily to manage, but only other alternative is to split out the internals so that recovery mode can use a logical plan creator that doesn't create topics. This is probably even messier.

Because of these, and the presence of an alternative (creating on REST mode), I'm going to close out these refactors.
